File: backend/src/services/bookingService.js
/**
 * Booking Service - Firebase Realtime Database
 */

const { v4: uuidv4 } = require('uuid');
const { getAll, getById, create, update, remove, queryByChild } = require('../config/firebase');

const BOOKINGS_REF = 'bookings';
const VENUES_REF = 'venues';
const USERS_REF = 'users';

/**
 * Enrich booking with venue and user data
 */
const enrichBooking = async (booking) => {
    if (!booking) return null;

    // Get venue data
    const venue = await getById(VENUES_REF, booking.venueId);
    // Get user data
    const user = await getById(USERS_REF, booking.userId);

    return {
        ...booking,
        venueName: venue?.name || null,
        venueType: venue?.type || null,
        venueAddress: venue?.address || null,
        userName: user?.name || null,
        userEmail: user?.email || null,
    };
};

/**
 * Get all bookings (admin)
 */
const getBookings = async () => {
    const bookings = await getAll(BOOKINGS_REF);
    // Sort by createdAt DESC
    bookings.sort((a, b) => new Date(b.createdAt) - new Date(a.createdAt));
    // Enrich with venue and user data
    return Promise.all(bookings.map(enrichBooking));
};

/**
 * Get bookings by user
 */
const getBookingsByUser = async (userId) => {
    const bookings = await queryByChild(BOOKINGS_REF, 'userId', userId);
    // Sort by createdAt DESC
    bookings.sort((a, b) => new Date(b.createdAt) - new Date(a.createdAt));
    // Enrich with venue data only
    return Promise.all(bookings.map(async (booking) => {
        const venue = await getById(VENUES_REF, booking.venueId);
        return {
            ...booking,
            venueName: venue?.name || null,
            venueType: venue?.type || null,
        };
    }));
};

/**
 * Get booking by ID
 */
const getBookingById = async (id) => {
    const booking = await getById(BOOKINGS_REF, id);
    return enrichBooking(booking);
};

/**
 * Create a new booking
 */
const createBooking = async ({ venueId, userId, bookingDate, startTime, endTime, totalPrice, notes }) => {
    const id = uuidv4();
    const now = new Date().toISOString();

    const bookingData = {
        venueId,
        userId,
        bookingDate,
        startTime,
        endTime,
        totalPrice: Number(totalPrice),
        status: 'pending',
        notes: notes || null,
        adminNotes: null,
        createdAt: now,
        updatedAt: now,
    };

    await create(BOOKINGS_REF, id, bookingData);
    return getBookingById(id);
};

/**
 * Update booking status (approve/reject)
 */
const updateBookingStatus = async (id, status, adminNotes = null) => {
    const updateData = {
        status,
        updatedAt: new Date().toISOString(),
    };

    if (adminNotes) {
        updateData.adminNotes = adminNotes;
    }

    await update(BOOKINGS_REF, id, updateData);
    return getBookingById(id);
};

/**
 * Cancel booking (by user)
 */
const cancelBooking = async (id, userId) => {
    const booking = await getById(BOOKINGS_REF, id);

    if (!booking) {
        throw new Error('Booking not found');
    }

    if (booking.userId !== userId) {
        throw new Error('Not authorized to cancel this booking');
    }

    if (booking.status !== 'pending') {
        throw new Error('Only pending bookings can be cancelled');
    }

    await update(BOOKINGS_REF, id, { status: 'cancelled', updatedAt: new Date().toISOString() });
    return getBookingById(id);
};

/**
 * Delete booking (admin)
 */
const deleteBooking = async (id) => {
    return remove(BOOKINGS_REF, id);
};

/**
 * Get pending bookings count
 */
const getPendingBookingsCount = async () => {
    const bookings = await queryByChild(BOOKINGS_REF, 'status', 'pending');
    return bookings.length;
};

module.exports = {
    getBookings,
    getBookingsByUser,
    getBookingById,
    createBooking,
    updateBookingStatus,
    cancelBooking,
    deleteBooking,
    getPendingBookingsCount,
};

File: frontend/lib/services/booking_service.dart
import 'api_client.dart';

class BookingService {
  BookingService._();

  static final BookingService instance = BookingService._();
  final ApiClient _client = ApiClient.instance;

  /// Get all bookings (admin)
  Future<List<BookingModel>> fetchAllBookings() async {
    final response = await _client.get('/bookings');
    final bookings = response['bookings'] as List<dynamic>? ?? [];
    return bookings
        .map((json) => BookingModel.fromJson(json as Map<String, dynamic>))
        .toList();
  }

  /// Get user's bookings
  Future<List<BookingModel>> fetchMyBookings() async {
    final response = await _client.get('/bookings/me');
    final bookings = response['bookings'] as List<dynamic>? ?? [];
    return bookings
        .map((json) => BookingModel.fromJson(json as Map<String, dynamic>))
        .toList();
  }

  /// Create booking
  Future<BookingModel> createBooking(Map<String, dynamic> data) async {
    final response = await _client.post('/bookings', data);
    return BookingModel.fromJson(response['booking'] as Map<String, dynamic>);
  }

  /// Update booking status (admin)
  Future<BookingModel> updateStatus(String id, String status, {String? adminNotes}) async {
    final response = await _client.patch('/bookings/$id/status', {
      'status': status,
      if (adminNotes != null) 'adminNotes': adminNotes,
    });
    return BookingModel.fromJson(response['booking'] as Map<String, dynamic>);
  }

  /// Cancel booking (user)
  Future<BookingModel> cancelBooking(String id) async {
    final response = await _client.post('/bookings/$id/cancel', {});
    return BookingModel.fromJson(response['booking'] as Map<String, dynamic>);
  }
}

class BookingModel {
  final String id;
  final String venueId;
  final String userId;
  final String bookingDate;
  final String startTime;
  final String endTime;
  final double totalPrice;
  final String status;
  final String? notes;
  final String? adminNotes;
  final String? venueName;
  final String? venueType;
  final String? userName;
  final String? userEmail;
  final DateTime? createdAt;

  BookingModel({
    required this.id,
    required this.venueId,
    required this.userId,
    required this.bookingDate,
    required this.startTime,
    required this.endTime,
    required this.totalPrice,
    required this.status,
    this.notes,
    this.adminNotes,
    this.venueName,
    this.venueType,
    this.userName,
    this.userEmail,
    this.createdAt,
  });

  factory BookingModel.fromJson(Map<String, dynamic> json) {
    return BookingModel(
      id: json['id'] as String? ?? '',
      venueId: json['venueId'] as String? ?? '',
      userId: json['userId'] as String? ?? '',
      bookingDate: json['bookingDate'] as String? ?? '',
      startTime: json['startTime'] as String? ?? '',
      endTime: json['endTime'] as String? ?? '',
      totalPrice: (json['totalPrice'] as num?)?.toDouble() ?? 0,
      status: json['status'] as String? ?? 'pending',
      notes: json['notes'] as String?,
      adminNotes: json['adminNotes'] as String?,
      venueName: json['venueName'] as String?,
      venueType: json['venueType'] as String?,
      userName: json['userName'] as String?,
      userEmail: json['userEmail'] as String?,
      createdAt: json['createdAt'] != null 
          ? DateTime.tryParse(json['createdAt'].toString()) 
          : null,
    );
  }
}
